deps/0000-metrics-health-check.md

```diff
@@ -0,0 +1,175 @@
+- No standardized health check mechanism for container orchestration systems (e.g., Kubernetes)
+- Limited visibility into component-level metrics and resource utilization
+- No centralized way to query and aggregate metrics/health across the distributed system
+
+## Goals
+
+* Implement a unified HTTP endpoint infrastructure for Dynamo components to expose metrics and health check endpoints
+* Enable customizable health checks through Python bindings while maintaining core health checks in Rust
+* Support standard observability patterns compatible with container orchestration systems
+
+
+## Requirements
+
+### REQ 1 Unified HTTP Endpoint Infrastructure
+
+The system **MUST** include a unified HTTP endpoint infrastructure for Dynamo components to expose metrics and health check endpoints
+
+### REQ 2 Performance Mrtics Requirements for worker nodes
+
+The metrics for worker nodes we want to monitor **MUST** include:
+- Inflight/Total Request: updated when a new request arrives (and finishes for inflight)
+- TTFT: reported at first chunk response
+- ITL: reported at each new chunk response
+- ISL: reported at first chunk response (TODO: report right after tokenization)
+- OSL: reported when requests finishes
+
+### REQ 3 Component Health Check Endpoints
+
+Each Dynamo component **MUST** expose HTTP endpoints for health monitoring:
+- `/health` - Overall component health status
+- `/liveness` - Component liveness probe
+- `/readiness` - Component readiness probe
+
+### REQ 4 Core Health Check Implementation
+
+The Rust runtime **MUST** implement basic health checks including:
+- etcd connectivity and lease status
+- NATS connectivity and service registration status
+
+### REQ 5 Extensible Health Check Framework
+
+The system **MUST** provide Python bindings that allow users to:
+- Register custom health check functions
+- Define component-specific health criteria
+- Customize health check response formats and thresholds
+
+
+
+# Architecture
+
+## Overview
+
+The proposed solution consists of three main components:
+
+1. **Unified HTTP Server Port**: Each Dynamo component will embed a single HTTP server that provides a unified interface for both metrics exposure and health check endpoints, eliminating the need for multiple ports or separate servers per component.
+
+2. **Metrics**: Component-level metrics collection and exposure through standardized HTTP endpoints, migrating from the existing approach implemented for Rust frontend where each component serves its own metrics data in standard formats (e.g., Prometheus).
+
+3. **Health Check**: Comprehensive health monitoring system with both Rust-implemented core health checks (etcd, NATS connectivity) and extensible Python-binding framework for custom health checks, exposed through standard HTTP endpoints compatible with container orchestration systems.
+
+## Unified HTTP Server Port
+
+To be done
+
+## Metrics Architecture
```

Comment:
How are metrics collected from within the component, e.g. if you need to move some metrics from the Python side? If I have a custom component that has custom metrics, how do I easily get those exposed?